comp/serializer/compression/impl-gzip/gzip_strategy.go

```diff
@@ -0,0 +1,139 @@
+}
+
+// NewComponent returns a new GzipStrategy
+func NewComponent(req Requires) Provides {
+	level := req.Level
+	if level < gzip.NoCompression {
+		log.Warnf("Gzip log level set to %d, minimum is %d.", level, gzip.NoCompression)
+		level = gzip.NoCompression
+	} else if level > gzip.BestCompression {
+		log.Warnf("Gzip log level set to %d, maximum is %d.", level, gzip.BestCompression)
+		level = gzip.BestCompression
+	}
+
+	return Provides{
+		Comp: &GzipStrategy{
+			level:   level,
+			newKind: req.NewKind,
+		},
+	}
+}
+
+// Compress will compress the data with gzip
+func (s *GzipStrategy) Compress(src []byte) (result []byte, err error) {
+	var compressedPayload bytes.Buffer
+	gzipWriter, err := gzip.NewWriterLevel(&compressedPayload, s.level)
+	defer func() {
+		err = gzipWriter.Close()
+		if err != nil {
+			result = nil
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	_, err = gzipWriter.Write(src)
+	if err != nil {
+		return nil, err
+	}
+	err = gzipWriter.Flush()
+	if err != nil {
+		return nil, err
+	}
+
+	return compressedPayload.Bytes(), nil
+}
+
+// Decompress will decompress the data with gzip
+func (s *GzipStrategy) Decompress(src []byte) ([]byte, error) {
+	reader, err := gzip.NewReader(bytes.NewReader(src))
+	if err != nil {
+		return nil, err
+	}
+	defer reader.Close()
+
+	// Read all decompressed data
+	var result bytes.Buffer
+	_, err = io.Copy(&result, reader)
+	if err != nil {
+		return nil, err
+	}
+
+	return result.Bytes(), nil
+}
+
+// CompressBound returns the worst case size needed for a destination buffer
+// when using gzip
+//
+// The worst case expansion is a few bytes for the gzip file header, plus
+// 5 bytes per 32 KiB block, or an expansion ratio of 0.015% for large files.
+// The additional 18 bytes comes from the header (10 bytes) and trailer
+// (8 bytes). There is no theoretical maximum to the header,
+// but we don't set any extra header fields so it is safe to assume
+//
+// Source: https://www.gnu.org/software/gzip/manual/html_node/Overview.html
+// More details are in the linked RFC: https://www.ietf.org/rfc/rfc1952.txt
+func (s *GzipStrategy) CompressBound(sourceLen int) int {
+	return sourceLen + (sourceLen/32768)*5 + 18
+}
+
+// ContentEncoding returns the content encoding value for gzip
+func (s *GzipStrategy) ContentEncoding() string {
+	return compression.GzipEncoding
+}
+
+// NewStreamCompressor returns a new gzip Writer
+func (s *GzipStrategy) NewStreamCompressor(output *bytes.Buffer) compression.StreamCompressor {
+	// Ensure level is within a range that doesn't cause NewWriterLevel to error.
+	level := s.level
+	if level < gzip.HuffmanOnly {
+		log.Warnf("Gzip streaming log level set to %d, minimum is %d.", level, gzip.HuffmanOnly)
+		level = gzip.HuffmanOnly
+	}
+
+	if level > gzip.BestCompression {
+		log.Warnf("Gzip streaming log level set to %d, maximum is %d.", level, gzip.BestCompression)
+		level = gzip.BestCompression
+	}
+
+	writer, _ := gzip.NewWriterLevel(output, level)
```

Comment:
We can consider grabbing the error to `log.Warnf` it here.

Reply:
The preconditions here mean that there will never be an error in practice, but I refactored this in c3f4e4e6fb since I think it is clearer.